### PR TITLE
Fix keyvault extension authentication to always use a user managed identity

### PR DESCRIPTION
**Pull request by tamirkamara**
## What is being addressed

The keyvault VM extension sometimes fails as it didn't have a config to always use the assigned user managed identity.

## How is this addressed

- Add config to set this.

### CODE REVIEW

**Comment by marrobi** [comment]
Ok, so tested this, bundle `2.5.1`.

Settings - off the VM itself:

```json
{"runtimeSettings": 
[{"handlerSettings": {"publicSettings": {"secretsManagementSettings": 
{"authenticationSettings": {"msiClientId": "/subscriptions/1e836626-b8c0-49a5-b4aa-9e9f93e31abf/resourceGroups/rg-mrtredemo27/providers/Microsoft.ManagedIdentity/userAssignedIdentities/id-nexus-mrtredemo27", 
"msiEndpoint": "http://169.254.169.254/metadata/identity"}, 
"observedCertificates": ["https://kv-mrtredemo27.vault.azure.net/secrets/nexusvmcert"], "pollingIntervalInS": "3600",
 "requireInitialSync": true}}, "protectedSettings": null, 
"protectedSettingsCertThumbprint": null}}]}
```

Extension logs:

```
2023-05-16 19:49:22: <debug> [Global]logFolder: /var/log/azure/Microsoft.Azure.KeyVault.KeyVaultForLinux
2023-05-16 19:49:22: <debug> [Global]Starting akvvm_service
2023-05-16 19:49:22: <info> [VMExtension]Starting extension
2023-05-16 19:49:22: <debug> [CertificateManagementConfiguration]Found v1.x configuration
2023-05-16 19:49:22: <info> [CertificateManagementConfiguration]Defaulting to MSI authentication.
2023-05-16 19:49:22: <info> [KVUnixService]Checking Linux distribution and version
2023-05-16 19:49:22: <info> [KVUnixService]OS info: ubuntu 18.04
2023-05-16 19:49:22: <info> [KVUnixService]Service Running...
2023-05-16 19:49:22: <info> [CertificateManager]Entering worker loop..
2023-05-16 19:49:22: <debug> [CertificateManager]MIN_POLLING_INTERVAL_SEC: 1, RandomMS: -1000
2023-05-16 19:49:22: <info> [CertificateManager]Starting refreshing observed certificates...
2023-05-16 19:49:22: <info> [CertificateManager]Beginning refresh for: https://kv-mrtredemo27.vault.azure.net/secrets/nexusvmcert
2023-05-16 19:49:22: <info> [KeyVaultClient]Getting new auth challenge
2023-05-16 19:49:22: <debug> [UnixKeyVaultHttpClient]Using CAfile for TLS: /etc/ssl/certs/ca-certificates.crt
2023-05-16 19:49:22: <debug> [UnixKeyVaultHttpClient]Using CApath for TLS: /etc/ssl/certs
2023-05-16 19:49:22: <debug> [AuthClient]AcquireTokenCallback invoked
2023-05-16 19:49:22: <debug> [AuthClient]acquiring token
2023-05-16 19:49:22: <debug> [MSIAuthClient]acquiring token via MSI
2023-05-16 19:49:22: <debug> [MSIHttpClient]MSI URL: http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-02-01&authority=https://l
ogin.microsoftonline.com/62d053bf-6309-4727-a3bf-bf604826c289&resource=https://vault.azure.net
2023-05-16 19:49:22: <debug> [UnixKeyVaultHttpClient]Using CAfile for TLS: /etc/ssl/certs/ca-certificates.crt
2023-05-16 19:49:22: <debug> [UnixKeyVaultHttpClient]Using CApath for TLS: /etc/ssl/certs
2023-05-16 19:49:22: <error> [CertificateManager]Refreshing 'https://kv-mrtredemo27.vault.azure.net/secrets/nexusvmcert' failed with RequestExcept
ion: 403; desc: {"error":{"code":"Forbidden","message":"The user, group or application 'appid=d1e82d40-40e3-4b22-a88b-b3d4ccfc0f48;oid=a6ed1cfb-b5b4-4d63
-940d-c414424b990e;iss=https://sts.windows.net/62d053bf-6309-4727-a3bf-bf604826c289/' does not have secrets get permission on key vault 'kv-mrtredemo27;l
ocation=westeurope'. For help resolving this issue, please see https://go.microsoft.com/fwlink/?linkid=2125287","innererror":{"code":"AccessDenied"}}}
2023-05-16 19:49:22: <error> [CertificateManager]Failed to download one or more certificates.
2023-05-16 19:49:22: <info> [UnixCertificateManager]Checking state of termination event with a timeout of 3599000
```

The id `a6ed1cfb-b5b4-4d63-940d-c414424b990e` is the VMs system assigned identity.

The cloud-init logs show:

```
Setting up Nexus SSL...
Checking for nexus-data/keystores directory...
Waiting for cert to be downloaded from KV...
ERROR - Timeout while waiting!
```

**Comment by tamirkamara** [comment]
@marrobi something is off here. I don't have any system managed identity on my nexus VM, nor is there one in the environment associated with this PR.
Are you sure you have the right bundle? I suggest maybe deleting the repo from the ACR and republishing (with a new version) to verify.

**Comment by marrobi** [comment]
@tamirkamara if I deploy a VM through the portal, nothing set, it gets a system ID, if I look at the ARM JSON used to deploy there isn't one set. So wonder if its a default somewhere?

**Comment by marrobi** [comment]
@tamirkamara I think the JSON in terraform isn't quite right:

https://learn.microsoft.com/en-us/azure/virtual-machines/extensions/key-vault-linux#extension-schema

 `authenticationSettings` isn't under `secretManagementSettings`, just under settings.
 
 Might try give it a go.

**Comment by marrobi** [comment]
Ok, with that change I get a different error:

```
2023-05-19 10:41:21: <debug> [MSIHttpClient]MSI URL: http://169.254.169.254/metadata/identity/oauth2/token?api-version=2018-02-01&authority=https://l
ogin.microsoftonline.com/62d053bf-6309-4727-a3bf-bf604826c289&resource=https://vault.azure.net&client_id=/subscriptions/1e836626-b8c0-49a5-b4aa-9e9f93e31
abf/resourceGroups/rg-mrtredemo27/providers/Microsoft.ManagedIdentity/userAssignedIdentities/id-nexus-mrtredemo27
2023-05-19 10:41:21: <error> [MSIAuthClient]failed to retrieve MSI token from response: {"error":"invalid_request","error_description":"Identity not 
found"}
2023-05-19 10:41:21: <error> [CertificateManager]Refreshing 'https://kv-mrtredemo27.vault.azure.net/secrets/nexusvmcert' failed with RequestExcept
ion: 400; desc: {"error":"invalid_request","error_description":"Identity not found"}
```

**Comment by marrobi** [comment]
The client ID was incorrect, it was `id`, rather that `client_id` in the terraform.

File: templates/shared_services/sonatype-nexus-vm/terraform/vm.tf [hunk 1]
```diff
@@ -215,7 +215,11 @@ resource "azurerm_virtual_machine_extension" "keyvault" {
...
       "observedCertificates" : [
         data.azurerm_key_vault_certificate.nexus_cert.versionless_secret_id
-      ]
+      ],
+      "authenticationSettings" : {
+        "msiEndpoint" : "http://169.254.169.254/metadata/identity",
+        "msiClientId" : azurerm_user_assigned_identity.nexus_msi.id
+      }
     }
   })
```
Comment (marrobi) [inline templates/shared_services/sonatype-nexus-vm/terraform/vm.tf:224]: Needs amending to:

```
  settings = jsonencode({
    "secretsManagementSettings" : {
      "pollingIntervalInS" : "3600",
      "requireInitialSync" : true,
      "observedCertificates" : [
        data.azurerm_key_vault_certificate.nexus_cert.versionless_secret_id
      ]
    },
    "authenticationSettings" : {
        "msiEndpoint" : "http://169.254.169.254/metadata/identity",
        "msiClientId" : azurerm_user_assigned_identity.nexus_msi.client_id
    }
  })
```

**Review by marrobi — CHANGES_REQUESTED** [review]
Need the change making as per comment.

**Review by marrobi — APPROVED** [review]
LGTM - thanks.

**Comment by tamirkamara** [comment]
/test-force-approve
nexus tests have passed.

**Pull request merged**